### PR TITLE
use tools.go pattern for managing linter

### PR DESCRIPTION
**Pull request by williambanfield**
This change adds a `tools.go` file.
Information on managing tools in this way can be found in the [modules wiki](https://github.com/golang/go/wiki/Modules#how-can-i-track-tool-dependencies-for-a-module). 

This pattern runs the linter version referenced in the module's `go.mod` The primary advantage that this pattern provides is that everyone will use the same version of the linter when running `make lint`.

An example of a project using this pattern can be found in the [go-fastly](https://github.com/fastly/go-fastly/blob/main/tools/tools.go) client.

### CODE REVIEW

**Review by alexanderbez — COMMENTED** [review]
Does this play nicely with our CI workflow? What exactly is the interplay between what CI runs and the make target?

File: go.mod [hunk 1]
```diff
@@ -11,13 +11,23 @@ require (
...
 	github.com/fortytw2/leaktest v1.3.0
 	github.com/go-kit/kit v0.10.0
+	github.com/go-lintpack/lintpack v0.5.2 // indirect
 	github.com/gogo/protobuf v1.3.2
 	github.com/golang/protobuf v1.5.2
+	github.com/golangci/errcheck v0.0.0-20181223084120-ef45e06d44b6 // indirect
+	github.com/golangci/go-tools v0.0.0-20190318055746-e32c54105b7c // indirect
+	github.com/golangci/goconst v0.0.0-20180610141641-041c5f2b40f3 // indirect
+	github.com/golangci/gocyclo v0.0.0-20180528134321-2becd97e67ee // indirect
+	github.com/golangci/golangci-lint v1.38.0 // indirect
```
Comment (tychoish) [inline go.mod:21]: I feel like this (now) isn't indirect, and I think the ideal situation is where we don't have the indirect dependencies in the go.mod file, but I don't think it really matters.

**Comment by williambanfield** [comment]
> Does this play nicely with our CI workflow? What exactly is the interplay between what CI runs and the make target?

Sadly, there is basically no interplay between CI and `make lint` at the moment. We use the [golangci-lint action](https://github.com/golangci/golangci-lint-action). This is described in the [.github/workflows/lint.yml](https://github.com/tendermint/tendermint/blob/69cf90c51e09a09de6d5c726db1b71842b0c647b/.github/workflows/lint.yml#L1). The github action should likely be preferred in CI as it will package the tool for delivery to our CI pipeline and provide nicer output without us having to do anything. 

This `tool.go` mechanism at least allows us to manage the version of the local golangci-lint version. Currently we only manage the version of the github action and pin it to version `1.38` of `golangci-lint`. This change pins the local version to be the same as the version used in CI. It's definitely not perfect but is likely an improvement over having the local tool be unversioned?

File: go.mod (hunk 1 above)
Comment (williambanfield) [reply go.mod:21]: This won't end up in our users' `go.mod` so I think it should be fine?

Comment (alexanderbez) [reply go.mod:21]: IMO is this OK 👍

**Comment by alexanderbez** [comment]
I see, that makes sense. Two things:

1. We should ensure the version we use in CI matches what we have declared in our deps.
2. This is a good alternative to running the linter in Docker which is the current solution we use in some projects.

Nice work 👍

**Review by alexanderbez — APPROVED** [review]

**Comment by marbar3778** [comment]
This is good. Can this be its own go.mod? I don't think it makes much of a difference, mainly curious

**Comment by williambanfield** [comment]
> This is good. Can this be its own go.mod? I don't think it makes much of a difference, mainly curious

Had to do some quick research on this, but it looks like yes!
From the output of `go help build`, we can specify a different modfile when using `go run`. The normal `go.mod` file will still be used to detect the module root but the alternate mod file can be used to track the deps. This seems like a nice change to make. 
```
        -modfile file
                in module aware mode, read (and possibly write) an alternate go.mod
                file instead of the one in the module root directory. A file named
                "go.mod" must still be present in order to determine the module root
                directory, but it is not accessed. When -modfile is specified, an
                alternate go.sum file is also used: its path is derived from the
                -modfile flag by trimming the ".mod" extension and appending ".sum".
```

**Pull request merged**